File: clients/desktop/Core/MozzServer.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Mozz.Desktop.Core;

/// <summary>
/// Signing in, mirroring a catalog, and turning a track into a playable URL.
///
/// A thin, typed layer over <see cref="MozzCore"/>'s command dispatcher. It
/// exists so the rest of the app never composes a request dictionary by hand and
/// never sees a magic command string — those belong next to the types they
/// produce, not scattered through view models.
///
/// It also owns the one thing the Swift core deliberately does not: persistence
/// of the auth token, via <see cref="ISecretStore"/>. The core hands a token out
/// of <c>Connect</c> and forgets it; this puts it in the platform's keystore and
/// hands it back on the next launch through <c>Attach</c>.
/// </summary>
public sealed class MozzServer(MozzCore core, ISecretStore secrets, string? accountsPath = null)
{
    private const string AccountsKey = "accounts.json";
    private readonly string _accountsPath = accountsPath ?? Path.Combine(AppPaths.SupportDirectory, AccountsKey);

    // MARK: Sign-in

    /// <summary>
    /// Jellyfin and Subsonic sign in with a username and password. Plex does not
    /// — see <see cref="BeginPlexLinkAsync"/>.
    /// </summary>
    public async Task<ServerAccount> ConnectAsync(
        BackendKind kind, string baseUrl, string username, string? password,
        string? apiKey = null, CancellationToken token = default)
    {
        var identifier = ClientIdentifier;
        var session = await core.CallAsync<SessionPayload>(new
        {
            cmd = "connect",
            kind = kind.Wire(),
            baseURL = baseUrl,
            username,
            password,
            apiKey,
            clientIdentifier = identifier,
        }, token).ConfigureAwait(false)
            ?? throw new MozzCoreException("The server returned no session");

        return Persist(session, username, identifier);
    }

    /// <summary>
    /// Start Plex's PIN flow. The user opens <see cref="PlexLink.LinkUrl"/> (or
    /// scans it), then <see cref="PollPlexLinkAsync"/> is called until it
    /// returns an account.
    /// </summary>
    public async Task<PlexLink> BeginPlexLinkAsync(CancellationToken token = default)
    {
        var pin = await core.CallAsync<PlexPinPayload>(new
        {
            cmd = "plexPin",
            clientIdentifier = ClientIdentifier,
        }, token).ConfigureAwait(false)
            ?? throw new MozzCoreException("Plex returned no PIN");

        return new PlexLink(pin.PinId, pin.Code, pin.ClientIdentifier, pin.LinkUrl);
    }

    /// <summary>
    /// Poll once. Returns null while the user has not finished linking, which is
    /// the normal case for the first several seconds.
    /// </summary>
    public async Task<ServerAccount?> PollPlexLinkAsync(PlexLink link, CancellationToken token = default)
    {
        var session = await core.CallAsync<SessionPayload>(new
        {
            cmd = "plexPinCheck",
            pinId = link.PinId,
            code = link.Code,
            clientIdentifier = link.ClientIdentifier,
        }, token).ConfigureAwait(false);

        // The core answers `{ "url": null }` for "not linked yet", which
        // deserializes to a session with no token.
        if (session is null || string.IsNullOrEmpty(session.Token)) return null;
        return Persist(session, username: null, link.ClientIdentifier);
    }

    // MARK: Attach

    /// <summary>
    /// Give the core the credentials for a saved account so sync and playback
    /// work. Called at launch for every saved account.
    /// </summary>
    public async Task AttachAsync(ServerAccount account, CancellationToken token = default)
    {
        var secret = secrets.Get(SecretKey(account.ServerId))
            ?? throw new MozzCoreException(
                $"No stored credential for {account.ServerName}. Sign in again.");

        await core.CallAsync<AttachPayload>(new
        {
            cmd = "attach",
            // The account's own id, not one derived from the address it happens
            // to be reachable at today. The catalogue, the likes and the play
            // history are keyed on this, and a server's address can change
            // without the server changing. See ADR-0017.
            serverId = account.ServerId,
            kind = account.Kind.Wire(),
            baseURL = account.BaseUrl,
            token = secret,
            userID = account.UserId,
            username = account.Username,
            serverName = account.ServerName,
            clientIdentifier = account.ClientIdentifier,
            musicSectionID = account.MusicSectionId,
        }, token).ConfigureAwait(false);
    }

    /// <summary>
    /// Attach an account and leave it ready to sync.
    ///
    /// Plex addresses its catalog by library section, and a newly signed-in
    /// account has none — the PIN flow yields an account and a server, never a
    /// section. Syncing in that state fails with "Plex music section not
    /// resolved". So resolve it once, save it, and re-attach, because the core
    /// builds its backend at attach time and the one built above still has no
    /// section. Returns the account to sync with, which may differ from the one
    /// passed in.
    /// </summary>
    public async Task<ServerAccount> AttachForSyncAsync(
        ServerAccount account, CancellationToken token = default)
    {
        await AttachAsync(account, token).ConfigureAwait(false);

        if (account.Kind != BackendKind.Plex || account.MusicSectionId is { Length: > 0 })
        {
            return account;
        }

        // Needs the attach above: `libraries` resolves against an attached
        // backend. For Plex it is musicSections(), which reads library/sections
        // directly and so does not itself need a section.
        var libraries = await LibrariesAsync(account.ServerId, token).ConfigureAwait(false);
        if (libraries is not { Count: > 0 })
        {
            throw new MozzCoreException(
                $"{account.ServerName} has no music library for Mozz to sync.");
        }

        var resolved = account with { MusicSectionId = libraries[0].Id };
        SaveAccount(resolved);
        await AttachAsync(resolved, token).ConfigureAwait(false);
        return resolved;
    }

    /// <summary>
    /// Ask the Plex account for an address that answers, and re-attach the
    /// account on it.
    ///
    /// Plex hands out several addresses for one server — LAN, remote, relay —
    /// and any of them can stop working while the others are fine. Signing in
    /// picks one and, left alone, keeps it forever: when it dies, every cover
    /// goes grey and playback fails, which reads as a broken app rather than a
    /// bad address. (One did die, for twenty minutes, on 2026-09-01. See
    /// ADR-0017.)
    ///
    /// The account's id is passed through unchanged, so the catalogue, the likes
    /// and the play history stay attached to it. Returns null when there was
    /// nothing to change — no account token, no candidate that answered, or the
    /// stored address already being the right one. That is not a failure worth
    /// surfacing: it means "still the best we know of".
    /// </summary>
    public async Task<ServerAccount?> RepointAccountAsync(
        ServerAccount account, CancellationToken token = default)
    {
        if (account.Kind != BackendKind.Plex) return null;
        var accountToken = secrets.Get($"plex.account.{account.ServerId}");
        if (accountToken is not { Length: > 0 }) return null;

        SessionPayload? session;
        try
        {
            session = await core.CallAsync<SessionPayload>(new
            {
                cmd = "plexResolve",
                serverId = account.ServerId,
                accountToken,
                machineIdentifier = account.MachineIdentifier,
                serverName = account.ServerName,
                clientIdentifier = account.ClientIdentifier,
            }, token).ConfigureAwait(false);
        }
        catch (MozzCoreException)
        {
            return null;
        }

        if (session is null
            || string.IsNullOrEmpty(session.BaseUrl)
            || string.IsNullOrEmpty(session.Token))
        {
            return null;
        }
        if (session.BaseUrl == account.BaseUrl
            && session.Token == secrets.Get(SecretKey(account.ServerId)))
        {
            return null;
        }

        var updated = account with
        {
            BaseUrl = session.BaseUrl,
            MachineIdentifier = session.MachineIdentifier ?? account.MachineIdentifier,
        };
        secrets.Set(SecretKey(updated.ServerId), session.Token);
        SaveAccount(updated);
        await AttachAsync(updated, token).ConfigureAwait(false);
        return updated;
    }

    public Task<IReadOnlyList<MusicLibrary>?> LibrariesAsync(
        string serverId, CancellationToken token = default)
        => core.CallAsync<IReadOnlyList<MusicLibrary>>(new { cmd = "libraries", serverId }, token);

    public async Task<ServerAccount> SelectMusicLibraryAsync(
        ServerAccount account,
        string libraryId,
        CancellationToken token = default)
    {
        if (string.IsNullOrWhiteSpace(libraryId)) return account;
        var updated = account with { MusicSectionId = libraryId };
        SaveAccount(updated);
        await AttachAsync(updated, token).ConfigureAwait(false);
        return updated;
    }

    // MARK: Sync

    public Task<SyncStart?> StartSyncAsync(string serverId, CancellationToken token = default)
        => core.CallAsync<SyncStart>(new { cmd = "sync", serverId }, token);

    public Task<SyncStatus?> SyncStatusAsync(string serverId, CancellationToken token = default)
        => core.CallAsync<SyncStatus>(new { cmd = "syncStatus", serverId }, token);

    /// <summary>
    /// Run a sync to completion, reporting progress. Polls rather than taking a
    /// callback because the C ABI is synchronous — see the note in
    /// <c>MozzSessionServer.swift</c>. Throws if the sync fails.
    /// </summary>
    public async Task<SyncStatus> SyncAsync(
        string serverId, IProgress<SyncStatus>? progress = null,
        CancellationToken token = default)
    {
        var start = await StartSyncAsync(serverId, token).ConfigureAwait(false);
        if (start is { Started: false, Reason: { } reason })
        {
            throw new MozzCoreException($"Sync did not start: {reason}");
        }

        while (true)
        {
            token.ThrowIfCancellationRequested();
            await Task.Delay(TimeSpan.FromMilliseconds(400), token).ConfigureAwait(false);

            var status = await SyncStatusAsync(serverId, token).ConfigureAwait(false);
            if (status is null) continue;
            progress?.Report(status);

            if (!status.Finished) continue;
            if (status.Error is { Length: > 0 } error) throw new MozzCoreException(error);
            return status;
        }
    }

    // MARK: Playback

    public Task<StreamSource?> StreamAsync(
        string serverId, string remoteId, int? maxBitrateKbps = null,
        CancellationToken token = default)
        => core.CallAsync<StreamSource>(new
        {
            cmd = "streamURL",
            serverId,
            remoteId,
            maxBitrateKbps,
        }, token);

    public async Task<string?> ArtworkUrlAsync(
        string serverId, string artworkKey, int size = 512, CancellationToken token = default)
    {
        var result = await core.CallAsync<UrlPayload>(new
        {
            cmd = "artworkURL", serverId, artworkKey, size,
        }, token).ConfigureAwait(false);
        return result?.Url;
    }

    // MARK: Saved accounts

    /// <summary>
    /// The accounts this installation knows about. Non-secret: the token itself
    /// is in the platform keystore, and this file holds only what is needed to
    /// look it up and to label the account in the UI.
    /// </summary>
    public IReadOnlyList<ServerAccount> SavedAccounts()
    {
        var path = _accountsPath;
        if (!File.Exists(path)) return [];
        try
        {
            return JsonSerializer.Deserialize<List<ServerAccount>>(File.ReadAllText(path)) ?? [];
        }
        catch (JsonException)
        {
            // A corrupt accounts file must not brick the app. The tokens are
            // still in the keystore; the worst case is signing in again.
            return [];
        }
    }

    public void ForgetAccount(string serverId)
    {
        var remaining = SavedAccounts().Where(a => a.ServerId != serverId).ToList();
        WriteAccounts(remaining);
        secrets.Set(SecretKey(serverId), null);
        secrets.Set($"plex.account.{serverId}", null);
    }

    public void ForgetAllAccounts()
    {
        foreach (var account in SavedAccounts())
        {
            secrets.Set(SecretKey(account.ServerId), null);
            secrets.Set($"plex.account.{account.ServerId}", null);
        }
        WriteAccounts([]);
    }

    private ServerAccount Persist(SessionPayload session, string? username, string identifier)
    {
        var account = new ServerAccount
        {
            ServerId = session.ServerId,
            Kind = BackendKindExtensions.Parse(session.Kind),
            BaseUrl = session.BaseUrl,
            ServerName = session.ServerName,
            UserId = session.UserId,
            Username = username,
            ClientIdentifier = string.IsNullOrEmpty(session.ClientIdentifier)
                ? identifier
                : session.ClientIdentifier,
            MusicSectionId = null,
            MachineIdentifier = session.MachineIdentifier,
        };

        secrets.Set(SecretKey(account.ServerId), session.Token);
        if (session.AccountToken is { Length: > 0 } accountToken)
        {
            // Plex's account token is distinct from the per-server access token
            // and is what re-discovers the account's other servers later.
            secrets.Set($"plex.account.{account.ServerId}", accountToken);
        }

        SaveAccount(account);
        return account;
    }

    /// <summary>Insert or replace one account, leaving the others alone.</summary>
    internal void SaveAccount(ServerAccount account)
    {
        var accounts = SavedAccounts().Where(a => a.ServerId != account.ServerId).ToList();
        accounts.Add(account);
        WriteAccounts(accounts);
    }

    internal void SaveAccount(ServerAccount account, string secret, string? accountToken = null)
    {
        secrets.Set(SecretKey(account.ServerId), secret);
        if (accountToken is not null) secrets.Set($"plex.account.{account.ServerId}", accountToken);
        SaveAccount(account);
    }

    private void WriteAccounts(IReadOnlyList<ServerAccount> accounts)
    {
        Directory.CreateDirectory(Path.GetDirectoryName(_accountsPath)!);
        File.WriteAllText(_accountsPath, JsonSerializer.Serialize(accounts,
            new JsonSerializerOptions { WriteIndented = true }));
    }

    private static string SecretKey(string serverId) => $"token.{serverId}";

    /// <summary>
    /// A stable per-installation id. Servers show it in their device list, so it
    /// must survive restarts — a new one each launch litters someone's Plex
    /// account with dozens of "Mozz on Windows" entries.
    /// </summary>
    private string ClientIdentifier
    {
        get
        {
            const string key = "clientIdentifier";
            var existing = secrets.Get(key);
            if (existing is { Length: > 0 }) return existing;
            var created = Guid.NewGuid().ToString();
            secrets.Set(key, created);
            return created;
        }
    }
}

// MARK: - Types

public enum BackendKind { Plex, Jellyfin, Subsonic }

public static class BackendKindExtensions
{
    public static string Wire(this BackendKind kind) => kind switch
    {
        BackendKind.Plex => "plex",
        BackendKind.Jellyfin => "jellyfin",
        BackendKind.Subsonic => "subsonic",
        _ => throw new ArgumentOutOfRangeException(nameof(kind)),
    };

    public static BackendKind Parse(string wire) => wire switch
    {
        "plex" => BackendKind.Plex,
        "jellyfin" => BackendKind.Jellyfin,
        "subsonic" => BackendKind.Subsonic,
        _ => throw new ArgumentOutOfRangeException(nameof(wire), wire, "unknown backend"),
    };

    public static string Display(this BackendKind kind) => kind switch
    {
        BackendKind.Plex => "Plex",
        BackendKind.Jellyfin => "Jellyfin",
        BackendKind.Subsonic => "Subsonic",
        _ => kind.ToString(),
    };
}

/// <summary>A saved server, minus its secret.</summary>
public sealed record ServerAccount
{
    public required string ServerId { get; init; }
    public required BackendKind Kind { get; init; }
    public required string BaseUrl { get; init; }
    public required string ServerName { get; init; }
    public string? UserId { get; init; }
    public string? Username { get; init; }
    public required string ClientIdentifier { get; init; }
    public string? MusicSectionId { get; init; }

    /// <summary>
    /// The Plex <em>server's</em> machine identifier — not this app's, which is
    /// <see cref="ClientIdentifier"/>.
    ///
    /// A Plex server has several addresses and none of them is the server; this
    /// is what stays the same when the address changes, so it is what
    /// re-resolution matches on. Null for accounts signed in before it was
    /// recorded.
    /// </summary>
    public string? MachineIdentifier { get; init; }
}

public sealed record PlexLink(int PinId, string Code, string ClientIdentifier, string? LinkUrl);

public sealed record MusicLibrary(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("name")] string Name);

public sealed record SyncStart(
    [property: JsonPropertyName("started")] bool Started,
    [property: JsonPropertyName("reason")] string? Reason);

public sealed record SyncStatus(
    [property: JsonPropertyName("running")] bool Running,
    [property: JsonPropertyName("finished")] bool Finished,
    [property: JsonPropertyName("phase")] string? Phase,
    [property: JsonPropertyName("itemsSynced")] int ItemsSynced,
    [property: JsonPropertyName("total")] int? Total,
    [property: JsonPropertyName("error")] string? Error,
    [property: JsonPropertyName("artists")] int? Artists,
    [property: JsonPropertyName("albums")] int? Albums,
    [property: JsonPropertyName("tracks")] int? Tracks,
    [property: JsonPropertyName("playlists")] int? Playlists)
{
    /// <summary>Short label for a progress row, e.g. "Songs 3,712 / 20,004".</summary>
    public string Describe()
    {
        var phase = Phase switch
        {
            "capabilities" => "Connecting",
            "syncing" => "Syncing",
            "artists" => "Artists",
            "albums" => "Albums",
            "tracks" => "Songs",
            "playlists" => "Playlists",
            "pruning" => "Finishing up",
            "done" => "Done",
            _ => "Syncing",
        };
        return Total is > 0
            ? $"{phase} {ItemsSynced:N0} / {Total:N0}"
            : $"{phase} {ItemsSynced:N0}";
    }
}

public sealed record StreamSource(
    [property: JsonPropertyName("url")] string Url,
    [property: JsonPropertyName("isTranscoded")] bool IsTranscoded,
    [property: JsonPropertyName("sessionID")] string? SessionId);

internal sealed record SessionPayload(
    [property: JsonPropertyName("serverId")] string ServerId,
    [property: JsonPropertyName("kind")] string Kind,
    [property: JsonPropertyName("baseURL")] string BaseUrl,
    [property: JsonPropertyName("token")] string Token,
    [property: JsonPropertyName("userID")] string? UserId,
    [property: JsonPropertyName("serverName")] string ServerName,
    [property: JsonPropertyName("clientIdentifier")] string ClientIdentifier,
    [property: JsonPropertyName("accountToken")] string? AccountToken,
    [property: JsonPropertyName("machineIdentifier")] string? MachineIdentifier = null);

internal sealed record PlexPinPayload(
    [property: JsonPropertyName("pinId")] int PinId,
    [property: JsonPropertyName("code")] string Code,
    [property: JsonPropertyName("clientIdentifier")] string ClientIdentifier,
    [property: JsonPropertyName("linkURL")] string? LinkUrl);

internal sealed record AttachPayload(
    [property: JsonPropertyName("serverId")] string ServerId);

internal sealed record UrlPayload(
    [property: JsonPropertyName("url")] string? Url);

File: clients/android/core/src/main/java/com/thatcube/mozz/core/MozzServer.kt
package com.thatcube.mozz.core

import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import java.io.File
import java.util.UUID

/**
 * Signing in, mirroring a catalog, and turning a track into a playable URL.
 *
 * A thin, typed layer over [MozzCore]'s command dispatcher. It exists so the
 * rest of the app never composes a request by hand and never sees a magic
 * command string — those belong next to the types they produce, not scattered
 * through view models.
 *
 * It also owns the one thing the Swift core deliberately does not: persistence
 * of the auth token, via [SecretStore]. The core hands a token out of `connect`
 * and forgets it; this puts it in the Keystore and hands it back on the next
 * launch through `attach`.
 *
 * The direct counterpart of `clients/desktop/Core/MozzServer.cs`, and kept
 * deliberately close to it — two clients diverging in how they drive the same
 * core is how subtle differences in behaviour get in.
 */
class MozzServer(
    private val core: MozzCore,
    private val secrets: SecretStore,
    private val accountsFile: File,
) {

    // MARK: Sign-in

    /**
     * Jellyfin and Subsonic sign in with a username and password. Plex does not
     * — see [beginPlexLink].
     */
    suspend fun connect(
        kind: BackendKind,
        baseUrl: String,
        username: String,
        password: String? = null,
        apiKey: String? = null,
    ): ServerAccount {
        val identifier = clientIdentifier()
        val session: SessionPayload = core.require(
            CoreRequest(
                cmd = "connect",
                kind = kind.wire,
                baseURL = baseUrl,
                username = username,
                password = password,
                apiKey = apiKey,
                clientIdentifier = identifier,
            )
        )
        if (session.serverId.isEmpty() || session.token.isEmpty()) {
            // SessionPayload defaults its fields so an unfinished Plex poll can
            // decode; a `connect` that comes back that way is a real failure and
            // must not be persisted as an account with no server.
            throw MozzCoreException("$baseUrl did not return a usable session.")
        }
        return persist(session, username, identifier)
    }

    /**
     * Start Plex's PIN flow. The user opens [PlexLink.linkUrl] in a browser and
     * approves there — no password is ever typed into Mozz — then
     * [pollPlexLink] is called until it returns an account.
     */
    suspend fun beginPlexLink(): PlexLink {
        val pin: PlexPinPayload = core.require(
            CoreRequest(cmd = "plexPin", clientIdentifier = clientIdentifier())
        )
        return PlexLink(pin.pinId, pin.code, pin.clientIdentifier, pin.linkURL)
    }

    /**
     * Poll once. Returns null while the user has not finished linking, which is
     * the normal case for the first several seconds.
     */
    suspend fun pollPlexLink(link: PlexLink): ServerAccount? {
        val session: SessionPayload = core.call(
            CoreRequest(
                cmd = "plexPinCheck",
                pinId = link.pinId,
                code = link.code,
                clientIdentifier = link.clientIdentifier,
            )
        ) ?: return null

        // The core answers `{"url": null}` — which decodes to all-defaults —
        // for "not linked yet". That is the normal case for the first several
        // seconds, so it is a null return rather than an error.
        if (session.token.isEmpty() || session.serverId.isEmpty()) return null
        return persist(session, username = null, identifier = link.clientIdentifier)
    }

    /**
     * Poll until the user finishes linking, or until [timeoutMillis] elapses.
     * Plex's PIN is short-lived, so this gives up rather than polling forever.
     *
     * Two things this has to survive, both learned the hard way:
     *
     * **Rate limiting.** Plex answers 429 if the PIN is checked too eagerly, and
     * it counts every client — a second device (or a stale session left polling)
     * spends the same budget. So a failed poll backs off exponentially instead of
     * hammering at a fixed interval.
     *
     * **Transient failures are not link failures.** A 429, or a dropped Wi-Fi
     * packet, means "ask again later", not "the user declined". Only the deadline
     * ends this loop unhappily; anything else is retried, and the last error is
     * reported if time runs out so the reason is not lost.
     */
    suspend fun awaitPlexLink(
        link: PlexLink,
        timeoutMillis: Long = 5 * 60 * 1000,
        pollMillis: Long = 2000,
        maxBackoffMillis: Long = 15000,
    ): ServerAccount {
        val deadline = System.currentTimeMillis() + timeoutMillis
        var wait = pollMillis
        var lastError: Throwable? = null

        while (System.currentTimeMillis() < deadline) {
            val outcome = runCatching { pollPlexLink(link) }
            outcome.getOrNull()?.let { return it }

            if (outcome.isSuccess) {
                // Not linked yet: the ordinary case. Keep the steady interval.
                wait = pollMillis
            } else {
                lastError = outcome.exceptionOrNull()
                wait = (wait * 2).coerceAtMost(maxBackoffMillis)
            }
            delay(wait)
        }

        throw MozzCoreException(
            lastError?.let { "Plex did not confirm the link: ${it.message}" }
                ?: "The Plex link expired before it was approved."
        )
    }

    // MARK: Attach

    /**
     * Give the core the credentials for a saved account so sync and playback
     * work. Called at launch for every saved account.
     */
    suspend fun attach(account: ServerAccount) {
        val secret = secrets.get(secretKey(account.serverId))
            ?: throw MozzCoreException(
                "No stored credential for ${account.serverName}. Sign in again."
            )

        core.call<Map<String, String>>(
            CoreRequest(
                cmd = "attach",
                kind = account.kind.wire,
                baseURL = account.baseUrl,
                token = secret,
                userID = account.userId,
                username = account.username,
                serverName = account.serverName,
                clientIdentifier = account.clientIdentifier,
                musicSectionID = account.musicSectionId,
            )
        )
    }

    /**
     * Attach an account and leave it ready to sync.
     *
     * Plex addresses its catalog by library section, and a newly linked account
     * has none — the PIN flow yields an account and a server, never a section.
     * Syncing in that state fails with "Plex music section not resolved". So
     * resolve it once, save it, and re-attach, because the core builds its
     * backend at attach time and the one built above still has no section.
     * Returns the account to sync with, which may differ from the one passed in.
     */
    suspend fun attachForSync(account: ServerAccount): ServerAccount {
        attach(account)

        if (account.kind != BackendKind.PLEX || !account.musicSectionId.isNullOrEmpty()) {
            return account
        }

        // Needs the attach above: `libraries` resolves against an attached
        // backend. For Plex it reads library/sections directly, so it does not
        // itself need a section.
        val libraries = libraries(account.serverId)
        if (libraries.isEmpty()) {
            throw MozzCoreException("${account.serverName} has no music library for Mozz to sync.")
        }

        val resolved = account.copy(musicSectionId = libraries.first().id)
        saveAccount(resolved)
        attach(resolved)
        return resolved
    }

    suspend fun libraries(serverId: String): List<MusicLibrary> =
        core.call<List<MusicLibrary>>(CoreRequest(cmd = "libraries", serverId = serverId))
            ?: emptyList()

    suspend fun selectMusicLibrary(account: ServerAccount, libraryId: String): ServerAccount {
        if (libraryId.isBlank()) return account
        val updated = account.copy(musicSectionId = libraryId)
        saveAccount(updated)
        attach(updated)
        return updated
    }

    // MARK: Sync

    /**
     * Run a sync to completion, emitting progress as it goes.
     *
     * Polls rather than taking a callback because the C ABI is synchronous —
     * there is no way for the core to call back into Kotlin mid-command. The
     * flow completes when the sync finishes and throws if it fails.
     */
    fun sync(serverId: String, pollMillis: Long = 400): Flow<SyncStatus> = flow {
        val start: SyncStart? = core.call(CoreRequest(cmd = "sync", serverId = serverId))
        if (start != null && !start.started) {
            throw MozzCoreException("Sync did not start: ${start.reason ?: "no reason given"}")
        }

        while (true) {
            delay(pollMillis)
            val status: SyncStatus = core.call(
                CoreRequest(cmd = "syncStatus", serverId = serverId)
            ) ?: continue

            emit(status)
            if (!status.finished) continue
            status.error?.takeIf { it.isNotEmpty() }?.let { throw MozzCoreException(it) }
            return@flow
        }
    }

    // MARK: Playback

    suspend fun stream(
        serverId: String,
        remoteId: String,
        maxBitrateKbps: Int? = null,
    ): StreamSource = core.require(
        CoreRequest(
            cmd = "streamURL",
            serverId = serverId,
            remoteId = remoteId,
            maxBitrateKbps = maxBitrateKbps,
        )
    )

    suspend fun artworkUrl(serverId: String, artworkKey: String, size: Int = 512): String? =
        core.call<UrlPayload>(
            CoreRequest(
                cmd = "artworkURL",
                serverId = serverId,
                artworkKey = artworkKey,
                size = size,
            )
        )?.url

    /**
     * What the attached server can do.
     *
     * Worth asking rather than inferring: the like control differs per backend —
     * a heart on Jellyfin, a star on Plex, both on Subsonic — and a client that
     * guessed from the backend's name would be wrong the first time a server grew
     * a feature. The core fetches this once per session and remembers it.
     */
    suspend fun capabilities(serverId: String): ServerCapabilities? {
        val fresh = core.call<ServerCapabilities>(
            CoreRequest(cmd = "capabilities", serverId = serverId)
        )
        if (fresh != null) rememberCapabilities(serverId, fresh)
        return fresh
    }

    /**
     * What this server could do last time it was asked.
     *
     * Answering this needs a network round trip, and until it lands a client has
     * to either show nothing or guess. Guessing is worse than waiting: a heart
     * that turns into a star a moment later has told the user something false
     * about their own library. Remembering the last answer means the question is
     * only ever open on the very first run.
     *
     * Deliberately separate from the account file. Capabilities are a fact about
     * the server that we cache, not part of the identity we signed in with, and
     * losing this file costs one round trip rather than an account.
     */
    fun cachedCapabilities(serverId: String): ServerCapabilities? {
        if (!capabilitiesFile.exists()) return null
        return try {
            MozzCore.json
                .decodeFromString<Map<String, ServerCapabilities>>(capabilitiesFile.readText())[serverId]
        } catch (error: Exception) {
            null
        }
    }

    private fun rememberCapabilities(serverId: String, value: ServerCapabilities) {
        val existing = try {
            if (capabilitiesFile.exists()) {
                MozzCore.json.decodeFromString<Map<String, ServerCapabilities>>(
                    capabilitiesFile.readText()
                )
            } else {
                emptyMap()
            }
        } catch (error: Exception) {
            emptyMap()
        }
        runCatching {
            capabilitiesFile.writeText(
                MozzCore.json.encodeToString(existing + (serverId to value))
            )
        }
    }

    private val capabilitiesFile: java.io.File
        get() = java.io.File(accountsFile.parentFile, "capabilities.json")

    // MARK: Saved accounts

    /**
     * The accounts this installation knows about. Non-secret: the token itself
     * is in the Keystore, and this file holds only what is needed to look it up
     * and to label the account in the UI.
     */
    fun savedAccounts(): List<ServerAccount> {
        if (!accountsFile.exists()) return emptyList()
        return try {
            MozzCore.json.decodeFromString<List<ServerAccount>>(accountsFile.readText())
        } catch (error: Exception) {
            // A corrupt accounts file must not brick the app. The tokens are
            // still in the Keystore; the worst case is signing in again.
            emptyList()
        }
    }

    fun forgetAccount(serverId: String) {
        writeAccounts(savedAccounts().filterNot { it.serverId == serverId })
        secrets.set(secretKey(serverId), null)
        secrets.set(plexAccountKey(serverId), null)
    }

    fun forgetAllAccounts() {
        savedAccounts().forEach {
            secrets.set(secretKey(it.serverId), null)
            secrets.set(plexAccountKey(it.serverId), null)
        }
        writeAccounts(emptyList())
    }

    /** Insert or replace one account, leaving the others alone. */
    fun saveAccount(account: ServerAccount) {
        writeAccounts(savedAccounts().filterNot { it.serverId == account.serverId } + account)
    }

    private fun persist(
        session: SessionPayload,
        username: String?,
        identifier: String,
    ): ServerAccount {
        val account = ServerAccount(
            serverId = session.serverId,
            kind = BackendKind.parse(session.kind),
            baseUrl = session.baseURL,
            serverName = session.serverName,
            userId = session.userID,
            username = username,
            clientIdentifier = session.clientIdentifier.ifEmpty { identifier },
            musicSectionId = null,
        )

        secrets.set(secretKey(account.serverId), session.token)
        session.accountToken?.takeIf { it.isNotEmpty() }?.let {
            // Plex's account token is distinct from the per-server access token
            // and is what re-discovers the account's other servers later.
            secrets.set(plexAccountKey(account.serverId), it)
        }

        saveAccount(account)
        return account
    }

    private fun writeAccounts(accounts: List<ServerAccount>) {
        accountsFile.parentFile?.mkdirs()
        accountsFile.writeText(MozzCore.json.encodeToString(accounts))
    }

    /**
     * A stable per-installation id. Servers show it in their device list, so it
     * has to survive restarts — a new one each launch litters someone's Plex
     * account with dozens of "Mozz on Android" entries.
     */
    private fun clientIdentifier(): String {
        secrets.get(CLIENT_IDENTIFIER)?.takeIf { it.isNotEmpty() }?.let { return it }
        return UUID.randomUUID().toString().also { secrets.set(CLIENT_IDENTIFIER, it) }
    }

    private companion object {
        const val CLIENT_IDENTIFIER = "clientIdentifier"
        fun secretKey(serverId: String) = "token.$serverId"
        fun plexAccountKey(serverId: String) = "plex.account.$serverId"
    }
}
